### PR TITLE
fix(text-field): remove foundation.setValue call during onChange

### PR DESCRIPTION
**Pull request by moog16**
fixes https://github.com/material-components/material-components-web-react/issues/333

By removing the componentDidUpdate from the index.js file, we no longer call foundation.setValue every input change. MDC Web was not doing this, so this is the correct fix.

### CODE REVIEW

File: packages/text-field/index.js [hunk 1]
```diff
@@ -262,7 +263,7 @@ class TextField extends React.Component {
         key='text-field-container'
       >
         {leadingIcon ? this.renderIcon(leadingIcon) : null}
-        {this.renderInput()}
+        {this.foundation_ ? this.renderInput() : null}
```
Comment (moog16) [inline packages/text-field/index.js]: This actually fixes https://github.com/material-components/material-components-web-react/issues/303

File: packages/text-field/Input.js [hunk 2]
```diff
@@ -43,20 +40,44 @@ export default class Input extends React.Component {
...
 
-    if (this.props.id !== prevProps.id) {
-      this.props.setInputId(this.props.id);
+    if (id !== prevProps.id) {
+      setInputId(id);
+    }
+
+    // this should be in the componentDidMount, but foundation is not created
+    // at that time.
+    if (value && foundation !== prevProps.foundation) {
```
Comment (moog16) [inline packages/text-field/Input.js]: Once this https://github.com/material-components/material-components-web-react/pull/353 is fixed, I will move this back into componentDidMount

File: packages/text-field/Input.js [hunk 3]
```diff
@@ -27,13 +27,10 @@ import classnames from 'classnames';
 import {VALIDATION_ATTR_WHITELIST} from '@material/textfield/constants';
 
 export default class Input extends React.Component {
-  constructor(props) {
-    super(props);
-    this.inputElement = React.createRef();
-  }
+  inputElement = React.createRef();
+  state = {disableProgramaticSetValue: false};
```
Comment (bonniezhou) [inline packages/text-field/Input.js]: Change this to `enableProgrammaticSetValue`, it's easier to reason about logic if the variable is positive

File: packages/text-field/Input.js [hunk 4]
```diff
@@ -96,7 +119,11 @@ export default class Input extends React.Component {
   // value of the input is.
   handleChange = (e) => {
     const {foundation, onChange} = this.props;
+    // autoCompleteFocus runs on `input` event in MDC Web. In React, onChange and
+    // onInput are the same event
+    // https://stackoverflow.com/questions/38256332/in-react-whats-the-difference-between-onchange-and-oninput
     foundation.autoCompleteFocus();
+    this.setState({disableProgramaticSetValue: true});
```
Comment (bonniezhou) [inline packages/text-field/Input.js]: So the `onChange` event is only emitted when it's from a user? Do programmatic changes emit any event?

File: packages/text-field/Input.js [hunk 5]
```diff
@@ -43,20 +40,46 @@ export default class Input extends React.Component {
...
-    if (this.props.value !== prevProps.value) {
-      this.props.handleValueChange(this.props.value);
+    if (value !== prevProps.value) {
+      handleValueChange(value, () => {
+        // only call #foundation.setValue on programatic changes;
+        // not changes by the user.
+        if (!this.state.disableProgramaticSetValue) {
+          foundation.setValue(value);
+        } else {
+          this.setState({disableProgramaticSetValue: false});
```
Comment (bonniezhou) [inline packages/text-field/Input.js]: at what point does the Text Field's `state.value` get updated if the change is from a user? should we be calling `handleValueChange` in `onBlur`?

File: packages/text-field/Input.js (hunk 4 above)
Comment (moog16) [reply packages/text-field/Input.js]: Right the onChange event is only triggered when a user types. BOTH events (user and programatic) trigger updates props, which calls `componentDidUpdate`.

File: packages/text-field/Input.js (hunk 5 above)
Comment (moog16) [reply packages/text-field/Input.js]: we call setValue in `handleBlur`, but its called within MDC's `foundation.deactivateFocus`

**Comment by moog16** [comment]
update - ready for re-review

File: packages/text-field/Input.js [hunk 6]
```diff
@@ -43,20 +40,46 @@ export default class Input extends React.Component {
...
-    if (this.props.id !== prevProps.id) {
-      this.props.setInputId(this.props.id);
+    if (id !== prevProps.id) {
+      setInputId(id);
+    }
+
+    // this should be in the componentDidMount, but foundation is not created
+    // at that time.
+    // Will fix this in
+    // https://github.com/material-components/material-components-web-react/pull/350/files
```
Comment (bonniezhou) [inline packages/text-field/Input.js]: should be 353, not 350

**Pull request merged**